pocs/tests/test_messaging.py

```diff
@@ -1,21 +1,41 @@
+import multiprocessing
 import pytest
 import time
 
 from datetime import datetime
-from multiprocessing import Process
 from pocs.utils.messaging import PanMessaging
 
 
 @pytest.fixture(scope='function')
 def forwarder():
-    def start_forwarder():
-        PanMessaging.create_forwarder(12345, 54321)
+    ready = multiprocessing.Event()
+    done = multiprocessing.Event()
 
-    messaging = Process(target=start_forwarder)
+    def start_forwarder():
+        PanMessaging.create_forwarder(
+            12345, 54321, ready_fn=lambda: ready.set(), done_fn=lambda: done.set())
 
+    messaging = multiprocessing.Process(target=start_forwarder)
     messaging.start()
+
+    if not ready.wait(timeout=10.0):
+        raise Exception('Forwarder failed to become ready!')
+    # Wait a moment for the forwarder to start using those sockets.
+    time.sleep(0.05)
+
     yield messaging
+
+    # Stop the forwarder. Since we use the same ports in multiple
+    # tests, we wait for the process to shutdown.
     messaging.terminate()
+    for _ in range(100):
+        # We can't be sure that the sub-process will succeed in
```

Comment:
Why can't we be sure of this?

Reply:
Because that is what the docs for multiprocessing.Process.terminate say.